src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java

```diff
@@ -0,0 +1,152 @@
+            processSelectedValues(selectedValues);
+        } else {
+            prcessSelectionAsGuid(input);
+        }
+    }
+
+    private void prcessSelectionAsGuid(String guid) throws CommCareSessionException {
+        ExternalDataInstance cachedInstance = virtualDataInstanceStorage.read(guid);
+        if (cachedInstance == null) {
+            throw new CommCareSessionException(
+                    "Could not make selection with reference id " + guid + " on this screen. " +
+                            " If this error persists please report a bug to CommCareHQ.");
+        }
+        storageReferenceId = guid;
+    }
+
+    private void processSelectedValues(String[] selectedValues)
+            throws CommCareSessionException {
+        if (selectedValues != null) {
+            String[] evaluatedValues = new String[selectedValues.length];
+            for (int i = 0; i < selectedValues.length; i++) {
+                TreeReference currentReference = getEntityReference(selectedValues[i]);
+                if (currentReference == null) {
+                    throw new CommCareSessionException(
+                            "Could not select case " + selectedValues[i] + " on this screen. " +
+                                    " If this error persists please report a bug to CommCareHQ.");
+                }
+                evaluatedValues[i] = getReturnValueFromSelection(currentReference);
+            }
+            ExternalDataInstance instance = VirtualInstances.buildSelectedValuesInstance(
+                    getSession().getNeededDatum().getDataId(),
+                    selectedValues);
+            String guid = virtualDataInstanceStorage.write(instance);
+            storageReferenceId = guid;
+
+            // rebuild instance with the source
+            ExternalDataInstanceSource instanceSource = buildSelectedValuesInstanceSource(instance,
+                    storageReferenceId);
+            selectedValuesInstance = instanceSource.toInstance();
+        }
+    }
+
+    @Override
+    protected void setSession(SessionWrapper session) throws CommCareSessionException {
+        super.setSession(session);
+        maxSelectValue = ((MultiSelectEntityDatum)mNeededDatum).getMaxSelectValue();
+    }
+
+    @Override
+    protected void updateSession(CommCareSession session) {
+        if (executePendingAction(session)) {
+            return;
+        }
+        if (storageReferenceId != null) {
+            if (selectedValuesInstance == null) {
+                selectedValuesInstance = virtualDataInstanceStorage.read(storageReferenceId);
+            }
+            ExternalDataInstanceSource externalDataInstanceSource = buildSelectedValuesInstanceSource(
+                    selectedValuesInstance, storageReferenceId);
+            session.setDatum(STATE_MULTIPLE_DATUM_VAL, mNeededDatum.getDataId(),
+                    storageReferenceId.toString(), externalDataInstanceSource);
```

Comment:
nit: storageReferenceId is already a string
```suggestion
                    storageReferenceId, externalDataInstanceSource);
```